package/kernel/linux/modules/lib.mk

```diff
@@ -7,6 +7,79 @@
 
 LIB_MENU:=Libraries
 
+define KernelPackage/lib-crypto-blake2s
+  SUBMENU:=$(LIB_MENU)
+  TITLE:=BLAKE2s hash function library
+  KCONFIG:=CONFIG_CRYPTO_LIB_BLAKE2S
+  FILES_PLATFORM:=$(LINUX_DIR)/lib/crypto/libblake2s-generic.ko
+  FILES:=$(LINUX_DIR)/lib/crypto/libblake2s.ko $(FILES_PLATFORM)
+endef
+
+define KernelPackage/lib-crypto-blake2s/x86/64
+  KCONFIG+=CONFIG_CRYPTO_BLAKE2S_X86
+  FILES_PLATFORM:=$(LINUX_DIR)/arch/x86/crypto/blake2s-x86_64.ko
+endef
+
+$(eval $(call KernelPackage,lib-crypto-blake2s))
+
+
+define KernelPackage/lib-crypto-chacha20poly1305
+  SUBMENU:=$(LIB_MENU)
+  TITLE:=ChaCha20-Poly1305 AEAD support (8-byte nonce library version)
+  KCONFIG:=CONFIG_CRYPTO_LIB_CHACHA20POLY1305
+  FILES_PLATFORM:= \
+	$(LINUX_DIR)/lib/crypto/libpoly1305.ko \
+	$(LINUX_DIR)/lib/crypto/libchacha.ko
+  FILES:=$(LINUX_DIR)/lib/crypto/libchacha20poly1305.ko $(FILES_PLATFORM)
+endef
+
+define KernelPackage/lib-crypto-chacha20poly1305/x86/64
+  KCONFIG+= \
+	CONFIG_CRYPTO_CHACHA20_X86_64 \
+	CONFIG_CRYPTO_POLY1305_X86_64
+  FILES_PLATFORM:= \
+	$(LINUX_DIR)/arch/x86/crypto/poly1305-x86_64.ko \
+	$(LINUX_DIR)/arch/x86/crypto/chacha-x86_64.ko
+endef
+
+define KernelPackage/lib-crypto-chacha20poly1305/arm
+  KCONFIG+=CONFIG_CRYPTO_POLY1305_ARM
+  FILES_PLATFORM:=$(LINUX_DIR)/arch/arm/crypto/poly1305-arm.ko
+endef
+
+KernelPackage/lib-crypto-chacha20poly1305/kirkwood=$(KernelPackage/lib-crypto-chacha20poly1305/arm)
+
+define KernelPackage/lib-crypto-chacha20poly1305/arm-neon
+  $(call KernelPackage/lib-crypto-chacha20poly1305/arm)
+  KCONFIG+=CONFIG_CRYPTO_CHAHA20_NEON
+  FILES_PLATFORM+=$(LINUX_DIR)/arch/arm/crypto/chacha-neon.ko
+endef
+
+KernelPackage/lib-crypto-chacha20poly1305/imx6=$(KernelPackage/lib-crypto-chacha20poly1305/arm-neon)
+KernelPackage/lib-crypto-chacha20poly1305/ipq40xx=$(KernelPackage/lib-crypto-chacha20poly1305/arm-neon)
+KernelPackage/lib-crypto-chacha20poly1305/mvebu/cortexa9=$(KernelPackage/lib-crypto-chacha20poly1305/arm-neon)
+
+define KernelPackage/lib-crypto-chacha20poly1305/arm64-neon
+  $(call KernelPackage/lib-crypto-chacha20poly1305/arm-neon)
```

Comment:
Why is the arm64 one calling the arm32 one?

Reply:
probably shouldn't if they are mutually exclusive. arm64 is not a superset of arm, is it? I'll fix it then

Reply:
Best to treat them like totally different architectures for our purposes here.